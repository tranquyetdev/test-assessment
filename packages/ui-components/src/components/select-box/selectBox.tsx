import { SelectBoxProps } from './types';
import React, { LegacyRef } from 'react';

// eslint-disable-next-line react/display-name
export const SelectBox = React.forwardRef((props: SelectBoxProps, ref: LegacyRef<HTMLSelectElement>) => {
  const {
    options,
    defaultValue,
    onChange,
    label,
    rightIcon,
    placeholder,
    disabled,
    className,
    size,
    variant,
    ..._props
  } = props;
  
  const [value, setValue] = React.useState(defaultValue);
  const styles = useStyles(props);

  return (
    <div className={`${styles.root} ${className ? className : ''}`}>
      {label && (
        <div
          className={`text-neutral-placeholder text-13 font-medium ${styles.label}`}
        >
          {label}
        </div>
      )}
      <div className={`relative w-full`}>
        <select
          {..._props}
          ref={ref}
          className={`
            w-full appearance-none focus:outline-none cursor-pointer 
            pl-[12px] pr-[48px] font-normal text-15
            ${styles.select}
            ${
              disabled
                ? 'border border-neutral-disable bg-neutral-disable text-neutral-placeholder'
                : `
                  border hover:border-neutral-placeholder 
                  focus:border-primary-base focus:border-[2px] focus:bg-neutral-white focus:text-neutral-text-primary
                  ${
                    value
                      ? 'border-neutral-text-primary text-neutral-primary bg-neutral-white'
                      : 'border-neutral-border bg-neutral-background text-neutral-placeholder'
                  }
                `
            }
          `}
          onChange = { (e) => {
            onChange?.(e);
            setValue(e.target.value) 
          }}
          defaultValue={defaultValue}
          disabled={disabled}
        >
          {placeholder && <option value={''}>{placeholder}</option>}
          {options?.map((option) => (
            <option key={option.value} value={option.value}>
              {option.name}
            </option>
          ))}
        </select>
        {rightIcon && (
          <div className="pointer-events-none absolute right-0 top-0 bottom-0 flex items-center px-2">
            {rightIcon}
          </div>
        )}
      </div>
    </div>
  );
});

const useStyles = ({ size, variant }: SelectBoxProps) => {
  return {
    root: variant === 'horizontal-label' ? 'flex items-center' : '',
    label: variant === 'horizontal-label' ? 'mr-[8px]' : 'mb-[8px]',
    select: `
      ${size === 'small' ? 'py-[4px]' : ''} 
      ${size === 'medium' ? 'py-[8px]' : ''}
      ${size === 'large' ? 'py-[12px]' : ''}
    `,
  };
};
