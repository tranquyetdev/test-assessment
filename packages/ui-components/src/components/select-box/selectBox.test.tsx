import React from 'react';
import { render } from '@testing-library/react';
import { SelectBox } from './selectBox';

describe('SelectBox', () => {
  Object.entries({}).forEach(([type, props]) => {
    it(`should render ${type} successfully`, () => {
      const { baseElement } = render(
        <SelectBox
          options={[]}
          onChange={function (value: string): void {
            throw new Error('Function not implemented.');
          }}
          label={''}
          variant={'verticle-label'}
          size={'small'}
        />
      );
      expect(baseElement).toMatchSnapshot();
    });
  });
});
