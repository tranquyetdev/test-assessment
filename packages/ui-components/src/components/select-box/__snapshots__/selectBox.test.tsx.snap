// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectBox should render SelectBox successfully 1`] = `
<body>
  <div>
    <div
      class=" undefined"
    >
      <div
        class="text-neutral-placeholder text-13 font-medium mb-[8px]"
      >
        Label
      </div>
      <div
        class="relative w-full"
      >
        <select
          class="
            w-full appearance-none focus:outline-none cursor-pointer 
            pl-[12px] pr-[48px] font-normal text-15
            
      py-[4px] 
      
      
    
            
                  border hover:border-neutral-placeholder 
                  focus:border-primary-base focus:border-[2px] focus:bg-neutral-white focus:text-neutral-text-primary
                  border-neutral-border bg-neutral-background text-neutral-placeholder
                
          "
        >
          <option
            value=""
          >
            Select your option
          </option>
          <option
            value="1"
          >
            option1
          </option>
          <option
            value="2"
          >
            option2
          </option>
          <option
            value="3"
          >
            option3
          </option>
          <option
            value="4"
          >
            option4
          </option>
          <option
            value="5"
          >
            option5
          </option>
        </select>
        <div
          class="pointer-events-none absolute right-0 top-0 bottom-0 flex items-center px-2"
        >
          <svg
            class="h-4 w-4"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</body>
`;
